sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -780,6 +817,217 @@ struct get_reduction_aux_kernel_name_t<sycl::detail::auto_name, Type, B1, B2,
+  size_t WGSize = NDRange.get_local_range().size();
+  size_t NWorkGroups = NDRange.get_group_range().size();
+
+  bool IsUpdateOfUserVar = !Reduction::is_usm && !Redu.initializeToIdentity();
+  auto PartialSums = Redu.getWriteAccForPartialReds(NWorkGroups, CGH);
+  auto Out =
+      (NWorkGroups == 1) ? PartialSums : Redu.getWriteAccForPartialReds(1, CGH);
+  auto LocalReds = Reduction::getReadWriteLocalAcc(WGSize + 1, CGH);
+  auto NWorkGroupsFinished =
+      Redu.getReadWriteAccessorToInitializedGroupsCounter(CGH);
+  auto DoReducePartialSumsInLastWG =
+      Reduction::template getReadWriteLocalAcc<int>(1, CGH);
+
+  auto Identity = Redu.getIdentity();
+  auto BOp = Redu.getBinaryOperation();
+  using Name =
+      typename get_reduction_main_kernel_name_t<KernelName, KernelType,
+                                                Reduction::is_usm, false>::name;
+  CGH.parallel_for<Name>(NDRange, [=](nd_item<1> NDId) {
+    // Call user's functions. Reducer.MValue gets initialized there.
+    typename Reduction::reducer_type Reducer(Identity, BOp);
+    reductionLoop(Range, Reducer, NDId, KernelFunc);
+
+    // Copy the element to local memory to prepare it for tree-reduction.
+    size_t LID = NDId.get_local_linear_id();
+    LocalReds[LID] = Reducer.MValue;
+    LocalReds[WGSize] = Identity;
+    DoReducePartialSumsInLastWG[0] = 0;
+    sycl::detail::workGroupBarrier();
+
+    // Tree-reduction: reduce the local array LocalReds[:] to LocalReds[0].
+    // LocalReds[WGSize] accumulates last/odd elements when the step
+    // of tree-reduction loop is not even.
+    size_t PrevStep = WGSize;
+    for (size_t CurStep = PrevStep >> 1; CurStep > 0; CurStep >>= 1) {
+      if (LID < CurStep)
+        LocalReds[LID] = BOp(LocalReds[LID], LocalReds[LID + CurStep]);
+      else if (LID == CurStep && (PrevStep & 0x1))
+        LocalReds[WGSize] = BOp(LocalReds[WGSize], LocalReds[PrevStep - 1]);
+      sycl::detail::workGroupBarrier();
+      PrevStep = CurStep;
+    }
+
+    if (LID == 0) {
+      auto V = BOp(LocalReds[0], LocalReds[WGSize]);
+      if (NWorkGroups == 1 && IsUpdateOfUserVar)
+        V = BOp(V, *Reduction::getOutPointer(Out));
+      // if NWorkGroups == 1, then PartialsSum and Out point to same memory.
+      Reduction::getOutPointer(PartialSums)[NDId.get_group_linear_id()] = V;
+
+      auto NFinished =
+          atomic_ref<int, memory_order::relaxed, memory_scope::device,
+                     access::address_space::global_space>(
+              NWorkGroupsFinished[0]);
+      if (++NFinished == NWorkGroups && NWorkGroups > 1)
+        DoReducePartialSumsInLastWG[0] = 1;
+    }
+
+    sycl::detail::workGroupBarrier();
+    if (DoReducePartialSumsInLastWG[0]) {
+      auto LocalSum = Identity;
+      for (size_t I = LID; I < NWorkGroups; I += WGSize)
+        LocalSum = BOp(LocalSum, Reduction::getOutPointer(PartialSums)[I]);
+      LocalReds[LID] = LocalSum;
+      LocalReds[WGSize] = Identity;
```

Comment:
Same comment as above.

Reply:
Fixed in 808bfa5122a6135f8e6feb45886cc0a59af18528